clang/lib/Serialization/ASTReaderDecl.cpp

```diff
@@ -617,24 +617,27 @@ void ASTDeclReader::VisitDecl(Decl *D) {
                            Reader.getContext());
   }
   D->setLocation(ThisDeclLoc);
-  D->InvalidDecl = Record.readInt();
-  if (Record.readInt()) { // hasAttrs
+
+  uint64_t DeclBits = Record.readInt();
+  D->InvalidDecl = DeclBits & 0x1;
+  D->setImplicit(DeclBits & (1 << 2));
+  D->Used = (DeclBits >> 3) & 0x1;
+  IsDeclMarkedUsed |= D->Used;
+  D->setReferenced(DeclBits & (1 << 4));
+  D->setTopLevelDeclInObjCContainer(DeclBits & (1 << 5));
+  D->setAccess((AccessSpecifier)((DeclBits >> 6) & 0x3));
+  D->FromASTFile = true;
+  auto ModuleOwnership = (Decl::ModuleOwnershipKind)((DeclBits >> 8) & 0x7);
```

Comment:
Is there a way to improve the API avoiding these magic constants? I suspect this would be difficult to read and maintain.

Reply:
What I had in mind is to introduce a BitPacker, but it can't avoid magic constants completely. e.g, we have to know that the width of `AccessSpecifier ` should be 2 and the width of `ModuleOwnershipKind` should be 3.

Then we can refactor the above section to:

```
BitPacker Bits = Record.readInt();
D->InvalidDecl = Bits.getNextBit();
D->setImplicit(Bits.getNextBit());
D->Used = Bits.getNextBit();
IsDeclMarkedUsed |= D->Used;
D->setReferenced(Bits.getNextBit());
D->setTopLevelDeclInObjCContainer(Bits.getNextBit());
D->setAccess((AccessSpecifier)Bits.getNextBit</*width = */2>()); // Still we can't avoid the magic constants completely.
D->FromASTFile = true;
auto ModuleOwnership = (Decl::ModuleOwnershipKind)Bits.getNextBit</*width = */3>());
```

Would you be happy about the above form?

Reply:
That looks more readable to me. In the case of `ModuleOwnershipKind` can we use the bit size either with `sizeof` or via `NextInContextAndBits`? If we use `sizeof` this could look like:
`auto ModuleOwnership = Bits.getNextBit<Decl::ModuleOwnershipKind>());` but we will likely read/write 8 bits instead of 3 in this case...

Reply:
Yes... and I feel the magic numbers for some types (e.g., `AccessSpecifier` and `ModuleOwnershipKind`) might not be so scaring. Since the devs can see `AccessSpecifier` and `ModuleOwnershipKind` here then they can get the size quickly. Also even if now we have to look at both ASTWriterDecl.cpp and ASTReaderDecl.cpp to change anything. I mean, it is already tightly bounded.

Reply:
Done